### ISSUE TITLE
GDMA is unstable in Mode 2

### ISSUE THREAD

**Issue by akumanatt** [gbdev/pandocs#299]
From my recent works in [GBC-NICCC](https://github.com/AYCEdemo/GBC-NICCC), which utilizes a carefully timed GDMA transfer during HBlank (dubbed HHDMA in the source code). I managed to squeeze up to 128 bytes per line without any problems. A transfer this large would make it end just right before mode 2 ends which should not be a problem since VRAM is also accessible in this mode. However, my friend started running it in SameBoy and noticed some failed transfer chunks. This was very weird so we ended up testing it on a real hardware. And oh boy...

https://user-images.githubusercontent.com/77432377/119998835-47f70680-bffb-11eb-8dd2-8792c76947e0.mp4

Reducing the transfer per line to 80 bytes fixes this (which ends before mode 2 starts). However, this behavior is quite weird since everything before 0:15 mark in that video was also performing 128 bytes per line transfer and it ends up fine, which I also leave it at that in the final ROM. The question is, what exact condition causes this to happen? And how did SameBoy, the only emulator I know so far, (barely) reproduce this?

**Comment by paulb-nl** [comment]
There are 2 rows of 10 sprites active there. Fetching a sprite takes 5 to 11 pixel cycles which lengthens mode 3 and shortens the following Hblank which makes the DMA transfer end well into mode 3.

**Comment by CasualPokePlayer** [comment]
I don't think that specific mode 1 shortening condition would be to blame in particular (any reasonably accurate emulator should catch that easily; except perhaps bgb), but that train of thought might be the source of the issues.

**Comment by akumanatt** [comment]
Alright, today I was reminded of this and decided to actually look into it and fix this. Turns out I actually forgot to clear on-screen sprites. And, according to Mesen-S, they will still render and lengthen the active display period even if they are set to not display in the LCDC register. Sure enough, this behavior turns out to be true in real hardware as the glitches go away once I fixed those.

So in conclusion, this issue is pretty much resolved now.
